src/storage/lock_manager/lock_wait_context.rs

```diff
@@ -53,6 +62,108 @@ pub struct LockWaitContextSharedState {
 
     /// The key on which lock waiting occurs.
     key: Key,
+
+    /// When a lock-waiting request (allow_lock_with_conflict == true) is
+    /// resumed, it's theoretically possible that the request meets lock
+    /// again, therefore it may need to be pushed to the lock waiting queue
+    /// again. Since the request is popped out from the queue when resuming
+    /// (which means the lock wait entry doesn't exist in the lock waiting
+    /// queue during the resumed execution), it's possible that timeout or
+    /// deadlock happens from `WaiterManager` during that time, which will
+    /// try to cancel the request. Therefore it leads to such a corner case:
+    ///
+    /// 1. (scheduler) A request enters lock waiting state, so an entry is
+    /// pushed to the `LockWaitQueues`, and a    message is sent to
+    /// `LockManager`. 2. (scheduler) After a while the entry is popped out
+    /// and resumed from the `LockWaitQueues`. 3. (scheduler) The request
+    /// resumes execution but still finds lock on the key.
+    ///     * This is possible to be caused by delayed-waking up or encountering
+    ///       error when writing a lock-releasing command to the engine.
+    /// 4. (lock_manager) At the same time, `LockManager` tries to cancel the
+    /// request due to timeout. But when    calling `finish_request`, the
+    /// entry cannot be found from the `LockWaitQueues`. So it    believes
+    /// that the entry is already popped out and resumed and does nothing.
+    /// 5. (scheduler) An entry is pushed to the `LockWaitQueues` due to
+    /// encountering lock at step 3. 6. Then the request becomes unable to
+    /// be canceled by timeout or other possible errors. In worst
+    ///    cases, the request may stuck in TiKV forever.
+    ///
+    /// To solve this problem, a `is_canceled` flag should be set when
+    /// `LockManager` tries to cancel it, before accessing the
+    /// `LockWaitQueues`; when an entry is pushed to the `LockWaitQueues`,
+    /// check if `is_canceled` is set after locking its inner map (ensures
+    /// exclusive access with `LockManager`), and if it's set, cancel the
+    /// request like how `LockManager` should have done.
```

Comment:
I think so. But I didn't find any solution to avoid managing the state in two sperated data structure. At least lock waiting queue and lock manager should be in a single module in the future.

Reply:
Maybe we could think about it in the future.